File: packages/swift-tex-core/Sources/TexCore/CompileError.swift
/// The failure category of a compile call, mirroring `tex_core_status`.
public enum CompileStatus: Sendable, Hashable {
    /// A caller contract violation (for example a null source with a
    /// nonzero length); unreachable through this module's safe entry points.
    case invalidArgument
    /// The source is not valid UTF-8.
    case invalidUTF8
    /// The source uses a construct outside the shipped surface.
    case unsupported
    /// The engine could not allocate memory.
    case allocationFailed
}

/// The structured fail-fast compile error (plan decision D4): a failed
/// compile produces no tree, only this value naming the offending bytes.
public struct CompileError: Error, Sendable, Hashable, CustomStringConvertible {
    /// The failure category.
    public let status: CompileStatus
    /// The source bytes the error names, when the failure has a location.
    public let range: SourceRange?
    /// The deterministic ASCII description, for example
    /// `unsupported command \frac`.
    public let message: String

    /// Creates an error value.
    public init(status: CompileStatus, range: SourceRange?, message: String) {
        self.status = status
        self.range = range
        self.message = message
    }

    /// The message, with the byte range appended when the error has one.
    public var description: String {
        if let range {
            return "\(message) (bytes \(range.begin)..\(range.end))"
        }
        return message
    }
}

File: packages/es-tex-core/src/compile-error.ts
import type { SourceRange } from "./model.js";

/** The failure category of a compile call, mirroring `tex_core_status`. */
export type CompileStatus = "invalidArgument" | "invalidUtf8" | "unsupported" | "allocationFailed";

/**
 * The structured fail-fast compile error: a failed
 * compile produces no tree, only this error naming the offending bytes.
 */
export class CompileError extends Error {
    /** The failure category. */
    readonly status: CompileStatus;
    /** The source bytes the error names, when the failure has a location. */
    readonly range: SourceRange | null;
    /** The deterministic ASCII description, e.g. `unsupported command \frac`. */
    readonly errorMessage: string;

    constructor(status: CompileStatus, range: SourceRange | null, errorMessage: string) {
        super(range ? `${errorMessage} (bytes ${range.begin}..${range.end})` : errorMessage);
        this.name = "CompileError";
        this.status = status;
        // Snapshot and freeze: the whole error is a frozen value, so its
        // range must not stay a live alias of a caller-mutable object.
        this.range = range === null ? null : Object.freeze({ begin: range.begin, end: range.end });
        this.errorMessage = errorMessage;
        Object.freeze(this);
    }
}
